ngraph/core/reference/include/ngraph/runtime/reference/einsum.hpp

```diff
@@ -0,0 +1,1210 @@
+                        new HostTensor(matmul_output->get_element_type(), back_shape));
+                    const AxisVector order = get_default_order(matmul_output->get_shape());
+                    ngraph::runtime::reference::reshape(
+                        reinterpret_cast<const char*>(matmul_output->get_data_ptr<T>()),
+                        reinterpret_cast<char*>(contract_output->get_data_ptr<T>()),
+                        matmul_output->get_shape(),
+                        order,
+                        back_shape,
+                        matmul_output->get_element_type().size());
+
+                    update_operands(inputs,
+                                    input_subscripts,
+                                    input_ind1,
+                                    input_ind2,
+                                    contract_output,
+                                    resultant_subscript);
+                }
+
+                template <typename T>
+                void einsum_impl(const HostTensorVector& inputs,
+                                 const HostTensorVector& outputs,
+                                 const std::string& equation)
+                {
+                    // parse equation to extract input subscripts and output subscript
+                    std::vector<std::string> input_subscripts;
+                    std::string output_subscript;
+                    ngraph::opset7::Einsum::parse_equation(
+                        equation, input_subscripts, output_subscript);
+
+                    // compute einsum path that is used to contract a pair of operands
+                    // in more optimal order
+                    size_t num_inputs = inputs.size();
+                    auto einsum_path = compute_einsum_path(num_inputs);
+
+                    HostTensorVector int_inputs = inputs;
+
+                    // contract inputs by Einsum until just one is remained
+                    for (auto const& inds_pair : einsum_path)
+                    {
+                        contract_two_inputs<T>(int_inputs,
+                                               input_subscripts,
+                                               output_subscript,
+                                               inds_pair.first,
+                                               inds_pair.second);
+                    }
+
+                    NGRAPH_CHECK(int_inputs.size() == 1);
+
+                    // extract diagonal for the single operand
+                    extract_diagonal<T>(int_inputs, input_subscripts, 0);
+
+                    // reduce dimensions for the remained input node
+                    reduce_input<T>(int_inputs, input_subscripts, output_subscript, 0);
+
+                    // transpose dimensions to layout required by the output subscript
+                    transpose_input<T>(int_inputs, input_subscripts, output_subscript, 0);
+
+                    // write the result into output host tensor
```

Comment:
nit: i think this comment is redundant, it's pretty clear from code below what it does

Reply:
Removed.